Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('heart_pandas.csv')

In [3]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex)**

In [4]:
## Ваш код
male_count = data[data['sex']=='male'].sex.count()

In [5]:
data.sex.count()

303

In [6]:
female_count = data['sex'].value_counts().female

In [7]:
## Ваш ответ:
print(f"Мужчин в датасете: {male_count}.")
print(f"Женщин в датасете: {female_count}.")

Мужчин в датасете: 207.
Женщин в датасете: 96.


 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента)**

In [8]:
## Ваш код
man_freq = data['sex'].value_counts().male / data.sex.count() * 100

In [9]:
## Ваш ответ:
print("%.2f %%" % man_freq)

68.32 %


**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца?**

In [10]:
## Ваш код
sick_male = data[((data['sex']=='male') & (data['target']== 1))].sex.count()
sick_female = data[((data['sex']=='female') & (data['target']== 1))].sex.count()

In [11]:
## Ваш ответ:
print(f"{sick_male} мужчин имеют заболевания сердца.")
print(f"{sick_female} женщин имеют заболевания сердца.")

93 мужчин имеют заболевания сердца.
72 женщин имеют заболевания сердца.


**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [12]:
## Ваш код
health_male = data[((data['sex']=='male') & (data['target']== 0))].sex.count()
total = data.sex.count()

In [13]:
## Ваш ответ:
print(f"{health_male / total}")

0.37623762376237624


**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [14]:
## Ваш код
sick_youngest = data[data['target']==1].age.min()

In [15]:
## Ваш ответ:
print(f"{sick_youngest} years")

29 years


**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [16]:
## Ваш код
health_oldest = data[data['target']==0].age.max()

In [17]:
## Ваш ответ:
print(f"{health_oldest} years")

77 years


**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [18]:
## Ваш код
youngest_sick_female = data[((data['sex']=='female') & (data['target']== 1))].age.min()

In [19]:
## Ваш ответ:
print(f"{youngest_sick_female} years")

34 years


**Какой средний возраст женщин?**

In [20]:
## Ваш код
avg_female_age = data[(data['sex']=='female')].age.mean()

In [21]:
## Ваш ответ:
print("%.2f" %avg_female_age)

55.68


**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [22]:
## Ваш код
data[data['target']== 1].age.mean()
data[data['target']== 1].age.std()

9.550650751946778

In [23]:
## Ваш ответ:
data[data['target']== 1].age.agg(['mean','std'])

mean    52.496970
std      9.550651
Name: age, dtype: float64

**Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol)**

In [24]:
## Ваш код
mean_chol = data.cholesterol.mean()
print(mean_chol)

246.26402640264027


In [25]:
(data[data['target']== 1].cholesterol < mean_chol).all()

False

In [26]:
## Ваш ответ:
if (data[data['target']== 1].cholesterol < mean_chol).all():
    print("Правда")
else:
    print("Неправда")

Неправда


**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [27]:
## Ваш код
data.groupby("rest_ecg").describe().T

rest_ecg                       ST-T wave abnormality  \
age                     count             152.000000   
                        mean               52.914474   
                        std                 9.229196   
                        min                34.000000   
                        25%                44.750000   
                        50%                53.500000   
                        75%                60.000000   
                        max                71.000000   
resting_blood_pressure  count             152.000000   
                        mean              129.065789   
                        std                16.423078   
                        min                94.000000   
                        25%               120.000000   
                        50%               128.500000   
                        75%               140.000000   
                        max               180.000000   
cholesterol             count             152.000000   
                        mean              237.269737   
                        std                47.050765   
                        min               126.000000   
                        25%               204.000000   
                        50%               232.000000   
                        75%               264.500000   
                        max               354.000000   
max_heart_rate_achieved count             152.000000   
                        mean              151.960526   
                        std                23.312414   
                        min                71.000000   
                        25%               139.750000   
                        50%               158.000000   
                        75%               169.250000   
                        max               194.000000   
st_depression           count             152.000000   
                        mean                0.879605   
                        std                 1.054561   
                        min                 0.000000   
                        25%                 0.000000   
                        50%                 0.550000   
                        75%                 1.400000   
                        max                 5.600000   
num_major_vessels       count             152.000000   
                        mean                0.638158   
                        std                 1.006925   
                        min                 0.000000   
                        25%                 0.000000   
                        50%                 0.000000   
                        75%                 1.000000   
                        max                 4.000000   
target                  count             152.000000   
                        mean                0.631579   
                        std                 0.483971   
                        min                 0.000000   
                        25%                 0.000000   
                        50%                 1.000000   
                        75%                 1.000000   
                        max                 1.000000   

rest_ecg                       left ventricular hypertrophy      normal  
age                     count                      4.000000  147.000000  
                        mean                      61.000000   55.687075  
                        std                       10.099505    8.675902  
                        min                       55.000000   29.000000  
                        25%                       55.000000   50.500000  
                        50%                       56.500000   57.000000  
                        75%                       62.500000   62.000000  
                        max                       76.000000   77.000000  
resting_blood_pressure  count                      4.000000  147.000000  
                        mean                     140.500000  134.02

In [28]:
## Ваш ответ: см. выше

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [29]:
## Ваш код
sick_male_depression = data[((data['sex']=='male') & (data['target']== 1))].st_depression.mean()
health_female_depression = data[((data['sex']=='female') & (data['target']== 0))].st_depression.mean()

In [30]:
## Ваш ответ:
if sick_male_depression > health_female_depression:
    print("Уровень депрессии выше у мужчин, страдающих болезнью сердца.")
elif sick_male_depression < health_female_depression:
    print("Уровень депрессии выше у женщин, не страдающих болезнью сердца.")
else:
    print("Уровень депрессии одинаковый.")
    

Уровень депрессии выше у женщин, не страдающих болезнью сердца.


**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [31]:
## Ваш код
lst = ["chest_pain_type", "rest_ecg", "thalassemia"]
for elem in lst:
    group = data.groupby([elem])
    print(group["cholesterol"].agg(['min', 'max']))

                  min  max
chest_pain_type           
atypical angina   126  564
non-anginal pain  182  298
typical angina    131  409
                              min  max
rest_ecg                              
ST-T wave abnormality         126  354
left ventricular hypertrophy  197  327
normal                        149  564
                   min  max
thalassemia                
fixed defect       141  417
normal             169  318
reversable defect  126  564


In [32]:
## Ваш ответ: см. выше

**Сколько значений может принимать каждый из категориальных признаков?**

In [33]:
## Ваш код
objects = []
for column in data.columns:
    if np.dtype(data[column]) == 'object':
        objects.append(column)
        print(f"{column}: {len(data[column].unique())}")

sex: 2
chest_pain_type: 3
fasting_blood_sugar: 2
rest_ecg: 3
exercise_induced_angina: 2
st_slope: 2
thalassemia: 3


In [34]:
## Ваш ответ: см. выше

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [35]:
## Ваш код
L = lambda x: x.value_counts().std()
data[objects].agg(L)

sex                         78.488853
chest_pain_type             85.860352
fasting_blood_sugar        150.613744
rest_ecg                    84.041656
exercise_induced_angina     74.246212
st_slope                    13.435029
thalassemia                 74.303432
dtype: float64

In [36]:
D = data[objects].agg(L)
D.argmax()

2

In [37]:
L = lambda x: x.value_counts().std()
D = data[objects].agg(L)

In [38]:
## Ваш ответ:
print(objects[D.argmax()])

fasting_blood_sugar


---